### DISCUSSION TITLE
comment injecter la clé Token avec CURL

### DISCUSSION THREAD

**Discussion by olivi25** [marian-m12l/studio#155] [Q&A]
Bonsoir,

je voudrai bien utiliser le Studio mais j'ai des problèmes pour la configuration.
Je suis en Firmware 2.6 (je pense j'ai fait la maj d'aout) V1 Chinoise.
J'ai configuré le fichier .bat (je suis sous W10) avec ajout : set FS_MOUNTPOINT=E (lecteur du lunii).
Par contre comment injecter le token? j'ai fait cela :
curl -o ~/.studio/db/official.json -v -X GET -H 'Accept: application/json' -H 'Content-Type: application/json; charset=UTF-8' -H 'X-AUTH-TOKEN: avec ici ma clé _TOKEN'_ https://server-data-prod.lunii.com/v2/packs
et j'ai des messages d'erreur :
curl: (6) Could not resolve host: application
Note: Unnecessary use of -X or --request, GET is already inferred.
* Could not resolve host: application
* Closing connection 1
curl: (6) Could not resolve host: application
Note: Unnecessary use of -X or --request, GET is already inferred.
* Rebuilt URL to: charset=UTF-8'/
* Could not resolve host: charset=UTF-8'
* Closing connection 2
curl: (6) Could not resolve host: charset=UTF-8'
Note: Unnecessary use of -X or --request, GET is already inferred.

et pleins d'autres merci pour l'aide

**Comment by agademer** [comment]
Sous Windows, je mets `set FS_MOUNTPOINT=E:\`, mais je ne sais pas si c'est nécessaire.

Pour le curl, mettez des guillemets (") à la place des apostrophe (') pour les paramètres et je crois que ça marchera mieux.

-H "X-AUTH-TOKEN: 0000000000000000000000000"

Bon courage ! Ça vaut la peine :)

**Reply by olivi25** [reply]
ça à l'aire de marche mais est ce que c’est normal que ce qui est dans mon lunii arrive comme cela?

![image](https://user-images.githubusercontent.com/77078295/103861629-056d7580-50be-11eb-8e43-07039133dfa3.png)

**Comment by vpoulailleau** [comment]
Pour ma part, je reçois un JSON tout vide, ce qui expliquerait le succès tout relatif de curl, et le mauvais affichage.

J'ai fait un script à partir de vos infos :

```sh
TOKEN="mon-super-long-token"
curl -o ~/.studio/db/official.json -v -X GET -H "Accept: application/json" -H "Content-Type: application/json; charset=UTF-8" -H "X-AUTH-TOKEN:$TOKEN" server-data-prod.lunii.com/v2/packs
```

Ce qui donne à l'exécution :

```
Note: Unnecessary use of -X or --request, GET is already inferred.
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0*   Trying 52.50.142.159...
* TCP_NODELAY set
* Connected to server-data-prod.lunii.com (52.50.142.159) port 80 (#0)
> GET /v2/packs HTTP/1.1
> Host: server-data-prod.lunii.com
> User-Agent: curl/7.58.0
> Accept: application/json
> Content-Type: application/json; charset=UTF-8
> X-AUTH-TOKEN:mon-fameux-token-du-script
> 
< HTTP/1.1 302 Found
< Server: Cowboy
< Connection: keep-alive
< X-Dns-Prefetch-Control: off
< X-Frame-Options: SAMEORIGIN
< Strict-Transport-Security: max-age=15552000; includeSubDomains
< X-Download-Options: noopen
< X-Content-Type-Options: nosniff
< X-Xss-Protection: 1; mode=block
< Location: https://server-data-prod.lunii.com/v2/packs
< Vary: Accept
< Content-Length: 0
< Date: Thu, 07 Jan 2021 07:17:50 GMT
< Via: 1.1 vegur
< 
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
* Connection #0 to host server-data-prod.lunii.com left intact
```

Je récupère donc un fichier vide. Une piste ?

**Comment by marian-m12l** [comment]
@olivi25 Que contient le fichier ?

@vpoulailleau la réponse HTTP est une simple redirection (statut 302), donc sans corps de message. Il doit y avoir une option dans curl pour suivre les redirections, ou plus simplement il faut indiquer directement dans la commande le protocole "https://"

J'ai l'impression qu'il y a pas mal de problèmes liés à cette étape, je vais faire une installation fraîche et vérifier si ça fonctionne toujours.

**Reply by vpoulailleau** [reply]
@marian-m12l Cela fonctionne correctement en mettant https dans l'url

Voici le script que je viens d'utiliser avec succès :

```sh
TOKEN="mon-super-long-token"
curl -o ~/.studio/db/official.json -v -X GET -H "Origin: https://server-data-prod.lunii.com" -H "Accept: application/json" -H "Content-Type: application/json; charset=UTF-8" -H "X-AUTH-TOKEN:$TOKEN" https://server-data-prod.lunii.com/v2/packs
```

![studio](https://user-images.githubusercontent.com/17438768/103864911-b0ccf900-50c3-11eb-8a2f-1f524009f53e.png)

PS : Dans trois jours, c'est l'anniversaire de ma fille Chloé, elle aura un nouveau pack d'histoire fait à la main avec STUdio !

**Comment by olivi25** [comment]
@vpoulailleau
La mienne demain, elle est déjà contente que j'ai pus transférer un pack.  
je vois que tu as un pack PATPATROUILLE, je vais devoir en faire un, j'ai une fan à la maison :)  

@marian-m12l  je peux transférer les histoires mais je n'ai pas d'image de ceux du luniistore
voici ce que cela donne quand je lance la commande : 

curl -o ~/.studio/db/official.json -v -X GET -H 'Accept: application/json' -H 'Content-Type: application/json; charset=UTF-8' -H 'X-AUTH-TOKEN: XXXXXXX' https://server-data-prod.lunii.com/v2/packs
Note: Unnecessary use of -X or --request, GET is already inferred.
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:--  0:00:01 --:--:--     0* Could not resolve host: application
* Closing connection 0
curl: (6) Could not resolve host: application
Note: Unnecessary use of -X or --request, GET is already inferred.
* Could not resolve host: application
* Closing connection 1
curl: (6) Could not resolve host: application
Note: Unnecessary use of -X or --request, GET is already inferred.
* Rebuilt URL to: charset=UTF-8'/
* Could not resolve host: charset=UTF-8'
* Closing connection 2
curl: (6) Could not resolve host: charset=UTF-8'
Note: Unnecessary use of -X or --request, GET is already inferred.
* Rebuilt URL to: XXXXXXX (c'est la clé TOKEN)'/
* Could not resolve host: eyJ**************************************************************************
* Closing connection 3
curl: (6) Could not resolve host: eyJ**************************************************************************
Note: Unnecessary use of -X or --request, GET is already inferred.
*   Trying 34.254.115.250...
* TCP_NODELAY set
* Connected to server-data-prod.lunii.com (34.254.115.250) port 443 (#4)
* schannel: SSL/TLS connection with server-data-prod.lunii.com port 443 (step 1/3)
* schannel: checking server certificate revocation
* schannel: sending initial handshake data: sending 197 bytes...
* schannel: sent initial handshake data: sent 197 bytes
* schannel: SSL/TLS connection with server-data-prod.lunii.com port 443 (step 2/3)
* schannel: failed to receive handshake, need more data
* schannel: SSL/TLS connection with server-data-prod.lunii.com port 443 (step 2/3)
* schannel: encrypted data got 3024
* schannel: encrypted data buffer: offset 3024 length 4096
* schannel: sending next handshake data: sending 126 bytes...
* schannel: SSL/TLS connection with server-data-prod.lunii.com port 443 (step 2/3)
* schannel: encrypted data got 51
* schannel: encrypted data buffer: offset 51 length 4096
* schannel: SSL/TLS handshake complete
* schannel: SSL/TLS connection with server-data-prod.lunii.com port 443 (step 3/3)
* schannel: stored credential handle in session cache
> GET /v2/packs HTTP/1.1
> Host: server-data-prod.lunii.com
> User-Agent: curl/7.55.1
> Accept: */*
>
* schannel: client wants to read 102400 bytes
* schannel: encdata_buffer resized 103424
* schannel: encrypted data buffer: offset 0 length 103424
* schannel: encrypted data got 521
* schannel: encrypted data buffer: offset 521 length 103424
* schannel: decrypted data length: 492
* schannel: decrypted data added: 492
* schannel: decrypted data cached: offset 492 length 102400
* schannel: encrypted data buffer: offset 0 length 103424
* schannel: decrypted data buffer: offset 492 length 102400
* schannel: schannel_recv cleanup
* schannel: decrypted data returned 492
* schannel: decrypted data buffer: offset 0 length 102400
< HTTP/1.1 401 Unauthorized
< Server: Cowboy
< Connection: keep-alive
< X-Dns-Prefetch-Control: off
< X-Frame-Options: SAMEORIGIN
< Strict-Transport-Security: max-age=15552000; includeSubDomains
< X-Download-Options: noopen
< X-Content-Type-Options: nosniff
< X-Xss-Protection: 1; mode=block
< Access-Control-Allow-Origin: *
< Content-Type: application/json; charset=utf-8
< Content-Length: 14
< Etag: W/"e-djIOxj5ZQwwTkpBrnkCZVeuVgbc"
< Date: Thu, 07 Jan 2021 16:58:20 GMT
< Via: 1.1 vegur
<
{"code":"1.1"}* Connection #4 to host server-data-prod.lunii.com left intact

**Reply by marian-m12l** [reply]
J'ai masqué le token, il contient un idenfiant.

**Comment by olivi25** [comment]
et pour ajouter dans .studio\db je n'ai aucun fichier official.json j'ai seulement unofficial.json

**Comment by MathieuIko** [comment]
@olivi25 La log d'exécution de `curl` semble indiquer que la ligne de commande est mal "parsée" dès le départ.
En effet, `curl` tente de se connecter sans succès à `application`, puis à `charset=UTF-8`, puis au token (comme s'il s'agissait d'une URL).
Enfin, `curl` se connecte à l'URL vers le serveur de Lunii, mais sans les entêtes HTTP qu'il faut (en particulier le token d'authentification), et reçoit en retour une erreur 401 (accès non autorisé).

La commande est-elle exécutée sur Linux ou Windows ? Les champs d'entête sont-ils entourés par des apostrophes `'` ou des guillemets `"` ?

**Reply by olivi25** [reply]
je suis sous Windows 10 64. 
j'ai testé apostrophes ' ou des guillemets " même problème je ne vois pas les images, c’est comme s'il voyait des packs dans le lunii mais impossible de savoir quoi.
je viens d’essayer avec guillemets voila le résultat

curl -o ~/.studio/db/official.json -v -X GET -H "Accept: application/json" -H "Content-Type: application/json; charset=UTF-8" -H "X-AUTH-TOKEN:XXXX" https://server-data-prod.lunii.com/v2/packs

Note: Unnecessary use of -X or --request, GET is already inferred.
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0*   Trying 52.51.120.129...
* TCP_NODELAY set
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0* Connected to server-data-prod.lunii.com (52.51.120.129) port 443 (#0)
* schannel: SSL/TLS connection with server-data-prod.lunii.com port 443 (step 1/3)
* schannel: checking server certificate revocation
* schannel: sending initial handshake data: sending 197 bytes...
* schannel: sent initial handshake data: sent 197 bytes
* schannel: SSL/TLS connection with server-data-prod.lunii.com port 443 (step 2/3)
* schannel: failed to receive handshake, need more data
* schannel: SSL/TLS connection with server-data-prod.lunii.com port 443 (step 2/3)
* schannel: encrypted data got 3024
* schannel: encrypted data buffer: offset 3024 length 4096
* schannel: sending next handshake data: sending 126 bytes...
* schannel: SSL/TLS connection with server-data-prod.lunii.com port 443 (step 2/3)
* schannel: encrypted data got 51
* schannel: encrypted data buffer: offset 51 length 4096
* schannel: SSL/TLS handshake complete
* schannel: SSL/TLS connection with server-data-prod.lunii.com port 443 (step 3/3)
* schannel: stored credential handle in session cache
> GET /v2/packs HTTP/1.1
> Host: server-data-prod.lunii.com
> User-Agent: curl/7.55.1
> Accept: application/json
> Content-Type: application/json; charset=UTF-8
> X-AUTH-TOKEN:XXXXXXX (c'est la clé TOKEN)
>
* schannel: client wants to read 102400 bytes
* schannel: encdata_buffer resized 103424
* schannel: encrypted data buffer: offset 0 length 103424
* schannel: encrypted data got 13175
* schannel: encrypted data buffer: offset 13175 length 103424
* schannel: decrypted data length: 476
* schannel: decrypted data added: 476
* schannel: decrypted data cached: offset 476 length 102400
* schannel: encrypted data length: 12670
* schannel: encrypted data cached: offset 12670 length 103424
* schannel: decrypted data length: 6801
* schannel: decrypted data added: 6801
* schannel: decrypted data cached: offset 7277 length 102400
* schannel: encrypted data length: 5840
* schannel: encrypted data cached: offset 5840 length 103424
* schannel: failed to decrypt data, need more data
* schannel: schannel_recv cleanup
* schannel: decrypted data returned 7277
* schannel: decrypted data buffer: offset 0 length 102400
< HTTP/1.1 200 OK
< Server: Cowboy
< Connection: keep-alive
< X-Dns-Prefetch-Control: off
< X-Frame-Options: SAMEORIGIN
< Strict-Transport-Security: max-age=15552000; includeSubDomains
< X-Download-Options: noopen
< X-Content-Type-Options: nosniff
< X-Xss-Protection: 1; mode=block
< Access-Control-Allow-Origin: *
< Content-Type: application/json; charset=utf-8
< Content-Length: 631318
< Etag: W/"9a216-cUd77loS17hEWGGPAcZ5ILs0uBc"
< Date: Thu, 07 Jan 2021 20:14:32 GMT
< Via: 1.1 vegur
<
{ [6801 bytes data]
Warning: Failed to create the file ~/.studio/db/official.json: No such file or
Warning: directory
* Failed writing body (0 != 6801)
  1  616k    1  6801    0     0   6801      0  0:01:32 --:--:--  0:01:32 10120
* Closing connection 0
* schannel: shutting down SSL/TLS connection with server-data-prod.lunii.com port 443
* schannel: clear security context handle
curl: (23) Failed writing body (0 != 6801)

![image](https://user-images.githubusercontent.com/77078295/103940879-5fa61f00-512e-11eb-8316-08b1d28e8850.png)

**Reply by marian-m12l** [reply]
> je suis sous Windows 10 64.
> ...

Le tilde fonctionne sur windows pour indiquer le répertoire home ?

**Reply by MathieuIko** [reply]
Sur Windows, le `~` n'a pas la signification particulière "dossier de l'utilisateur".
Il faut plutôt utiliser `%userprofile%` (en Invite de Commandes) ou `$env:userprofile` (en PowerShell), pour obtenir le chemin vers le dossier de l'utilisateur.

**Comment by marian-m12l** [comment]
@olivi25 Effectivement la commande ne semble pas bien comprise par curl. Peut-être bien un problème de `'` ou `"`, oui. Sur quel OS ? quel shell ? Le but de cette commande est justement de créer le fichier official.json contenant la réponse du serveur lunii.

**Reply by olivi25** [reply]
je lance la commande depuis un CMD

en powershell voila l'erreur

Invoke-WebRequest : Impossible de lier le paramètre «Headers». Impossible de convertir la valeur «Accept:
application/json» du type «System.String» en type «System.Collections.IDictionary».
Au caractère Ligne:1 : 49
+ ... udio/db/official.json -v -X GET -H "Accept: application/json" -H "Con ...
+                                        ~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidArgument : (:) [Invoke-WebRequest], ParameterBindingException
    + FullyQualifiedErrorId : CannotConvertArgumentNoMessage,Microsoft.PowerShell.Commands.InvokeWebRequestCommand

**Reply by marian-m12l** [reply]
Et en précisant le chemin absolu vers le fichier official.json ?

**Comment by olivi25** [comment]
Je viens de tester avec le chemin absolu c'est bon :)
Je pense à préciser dans la DOC.
Il manque juste ceux que j'ai fait moi avec l'application android (studio officiel).

**Reply by marian-m12l** [reply]
Pour les packs créés _via_ Mon Studio Lunii, c'est normal. Il y a une issue à ce sujet: #117

**Comment by olivi25** [comment]
Merci beaucoup pour votre aide,

je pense dernière question savez vous pourquoi certain pack qui fonctionne avec le firmware V1 et plus avec le Firmware 2.6.
Voici un exemple de pack, dès que je suis dessus il fait redémarrer la lunii

https://drive.google.com/file/d/1R-DgVKXiCCO6kARIi9GaJ7LPAwsu1mjd/view

**Reply by marian-m12l** [reply]
Le support du nouveau firmware n'est pas encore stable.
Je viens de pousser la conversion/réencodage à la volée des fichier audio en MP3, ça devrait améliorer la compatibilité des packs.

**Reply by olivi25** [reply]
Bonsoir marian-m12l 
SI j'ai bien compris, avant le transfert dans le lunii, il y a une conversion MP3 + autre et dans le 0.3 Alpha ce n'est pas encore intégré ou mal c'est cela?

Merci pour tout

**Reply by marian-m12l** [reply]
@olivi25 Oui, conversion des fichiers audio en MP3 (44100Hz / Mono) et des images en BMP (profondeur de couleur 4-bit, encodage RLE) au moment du transfert, si nécessaire. En principe la dernière nightly gère ces deux conversions, mais ça n'est peut-être pas très fiable.

**Reply by olivi25** [reply]
je vais attendre la prochaine alpha, les nightly sont déjà compilées comme les alpha?

**Reply by marian-m12l** [reply]
Les nightly sont compilées en continu par une action github : https://github.com/marian-m12l/studio/actions?query=workflow%3A%22Nightly+Build%22

Chaque build génère un artefact équivalent aux releases.

**Reply by padider** [reply]
Merci pour le lien
J'ai modif l'image qui était pas au bon format et je l'ai intégré cela marche bien
J'ai meme compléter les histoires macdo pour avoir les 12

Ma fille adore du coup je me suis mis aussi a en faire d'autre
C'est pas parfait mais c'est sympa je trouve

Charles_Perrault = > https://uptobox.com/43wju3yxfyvs
Disney_V1 = > https://uptobox.com/rvg893rlur23
McDo_-_En_avant_la_musique_V4 = > https://uptobox.com/64ua71rsh0ot

Si d autres ont des liens a échanger

**Comment by TeensyDucM** [comment]
pour synthetiser, j'ai reussi à créer le fichier official.json en utilisant cmd ( pas le powershell) et en specifiant le chemin absolu vers le dossier .studio/db et en utilisant des ".

Par contre celui-ci ne doit pas etre correct car pas de creation du STUdio . le json donne ceci

{Code1;1}

**Reply by marian-m12l** [reply]
@TeensyDucM Ce code d'erreur peut venir d'un token invalide.

**Comment by jyoci** [comment]
La commande curl ne fonctionne pas bien (je suis sous Windows 10 64.) si on met des espaces dans les paramètres, par exemple 'Accept: application/json' ==>Could not resolve host: application. il faut supprimer ces espaces: 'Accept:application/json'   'Content-Type:application/json;charset=UTF-8'
Par ailleurs, ça a fonctionné mais cela ne fonctionne plus, j'ai une erreur  HTTP/1.1 401 Unauthorized... alors que je pense mettre le bon token...?!?!?    @TeensyDucM ça donne effectivement {Code1;1} dans ce cas

**Reply by MathieuIko** [reply]
As-tu essayé la commande `curl` en remplaçant les apostrophes `'` par des guillemets `"` ?
Dans le terminal PowerShell de Windows, ça fait une différence sur la manière dont il décompose la commande et ses paramètres.

Je viens de tester la commande (sur Linux), avec un vieux token d'ailleurs, et j'ai bien reçu les metadatas des packs officiels.
À quoi ressemble la commande que tu exécutes ? Le token est-il bien celui récupéré dans le fichier `.local.properties` ?

**Reply by jyoci** [reply]
Oui j'ai essayé. Je lance sous CMD. Mon token vient bien de  .local.properties , j'ai essaye non connecté et connecté.
Aurais-tu une ligne de commande avec un (vieux ;o) token juste pour essayer?

**Reply by vpoulailleau** [reply]
Normalement, avec la dernière version (béta 3), c'est automatique, plus besoin de curl ou autre. Pour la récupérer, c'est ici : https://github.com/marian-m12l/studio/releases/tag/0.3.0-beta.3

**Reply by marian-m12l** [reply]
Effectivement c'est automatisé dans la version 0.3.0 :-)

**Comment by olivi25** [comment]
Voila la mienne modifie le token j'en ai enlevé
donc modifie entre TOKEN: et " https://server-data-prod.lunii.com/v2/packs
J'ai testé hier plus de problèmes
Pour moi la V3 c'est que tu ne rentre plus la lettre de ton lecteur (E:) dans le .bat de studio-windows.

curl -o c:/Users/utilisateur/.studio/db/official.json -v -X GET -H "Accept: application/json" -H "Content-Type: application/json; charset=UTF-8" -H "X-AUTH-TOKEN:eyJ0eXAiOahzl4paw" https://server-data-prod.lunii.com/v2/packs
